### PR TITLE
Vexflow 4.0 migration (includes webpack 5.0) 

### PR DESCRIPTION
**Pull request by rvilarl**
No default Vex, classes exported in vexflow.

### CODE REVIEW

**Comment by sschmidTU** [comment]
Great, we'll need to test vexflow 4.x, also with the visual regression tests.
But for any change that's not directly related to vexflow 4, please at least explain why you are doing it.
For example, you removed the webpack --colors option, not sure why.
Also, i'm not sure that upgrading every single package in our package.json that mostly doesn't have anything to do with vexflow is a good idea for this PR. I'd prefer to keep it clean.

**Comment by rvilarl** [comment]
> For example, you removed the webpack --colors option, not sure why. Also, i'm not sure that upgrading every single package in our package.json that mostly doesn't have anything to do with vexflow is a good idea for this PR. 

--colors is no longer an option in webpack 5. I upgraded to tray and resolve the issues, as you said it should be done anyway. :)

**Comment by rvilarl** [comment]
@sschmidTU Simon could you review the changes that I made? It would help me to identify further issues. The tests are failing due to a bug in VexFlow, I have set a PR already.

I have seen that you are not using `strict` in the project. With strict the current develop branch produces 1123 errors. This branch has reduced the number to 966. It is a good start!

**Comment by sschmidTU** [comment]
`strict` for what? webpack?

**Comment by rvilarl** [comment]
> `strict` for what? webpack?

for TypeScript, it is an option in `tsconfig.json` under `compilerOptions`
```
"strict": true /* Enable all strict type-checking options. */
```
it helps a lot to find bugs and typing problems. There are a lot of `any` for example in OSMD.

**Comment by sschmidTU** [comment]
Yeah, `any` was necessary with Vexflow in Javascript and inconsistent type definitions.

**Comment by rvilarl** [comment]
:) ok I wait for your feedback then.

File: src/MusicalScore/Graphical/VexFlow/VexFlowMeasure.ts [hunk 1]
```diff
@@ -1238,11 +1247,11 @@ export class VexFlowMeasure extends GraphicalMeasure {
                     const vfse: VexFlowStaffEntry = vexFlowVoiceEntry.parentStaffEntry as VexFlowStaffEntry;
                     if (vfse && vfse.vfClefBefore) {
                         // add clef as NoteSubGroup so that we get modifier layouting
-                        const clefModifier: NoteSubGroup = new NoteSubGroup( [vfse.vfClefBefore] );
+                        const clefModifier: NoteSubGroup = new Vex.Flow.NoteSubGroup( [vfse.vfClefBefore] );
```
Comment (rvilarl) [inline src/MusicalScore/Graphical/VexFlow/VexFlowMeasure.ts:1250]: Probably an error, should use the local `NoteSubGroup` and not the one from `vexflow`

Comment (sschmidTU) [reply src/MusicalScore/Graphical/VexFlow/VexFlowMeasure.ts:1250]: Yes, you're importing NoteSubGroup already, no need for `Vex.Flow`, it looks to me.

File: webpack.common.js [hunk 2]
```diff
@@ -46,7 +46,7 @@ module.exports = {
         })
     ],
     devServer: {
-        contentBase: [
+        static: [
```
Comment (rvilarl) [inline webpack.common.js:49]: contentBase was renamed to static in WebPack 5.x

File: webpack.dev.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
-var merge = require('webpack-merge')
+var { merge } = require('webpack-merge')
```
Comment (rvilarl) [inline webpack.dev.js:1]: Export sysntaxt changed for new version of webpack-merge.

File: webpack.prod.js [hunk 4]
```diff
@@ -1,8 +1,8 @@
-var merge = require('webpack-merge')
+const { merge } = require('webpack-merge');
 var webpack = require('webpack')
 var path = require('path')
 var common = require('./webpack.common.js')
-var Visualizer = require('webpack-visualizer-plugin')
+var Visualizer = require('webpack-visualizer-plugin2')
```
Comment (rvilarl) [inline webpack.prod.js:5]: webpack-visualizer-plugin no longer compatible.

File: package.json [hunk 5]
```diff
@@ -10,15 +10,14 @@
...
-    "build:webpack": "webpack --progress --colors --config webpack.prod.js",
-    "build:webpack-dev": "webpack --progress --colors --config webpack.dev.js",
-    "build:webpack-sourcemap": "webpack --progress --colors --config webpack.sourcemap.js",
-    "start": "webpack-dev-server --progress --colors --config webpack.dev.js",
-    "start:local": "webpack-dev-server --progress --colors --config webpack.local.js",
+    "build:webpack": "webpack --progress --config webpack.prod.js",
+    "build:webpack-dev": "webpack --progress --config webpack.dev.js",
+    "build:webpack-sourcemap": "webpack --progress --config webpack.sourcemap.js",
+    "start": "webpack-dev-server --progress --config webpack.dev.js",
+    "start:local": "webpack-dev-server --progress --config webpack.local.js",
```
Comment (rvilarl) [inline package.json:20]: --colors no longer supported by webpack 5.x.

File: src/MusicalScore/Graphical/MusicSheetDrawer.ts [hunk 6]
```diff
@@ -143,13 +143,13 @@ export abstract class MusicSheetDrawer {
         throw new Error("not implemented");
     }
 
-    public drawLabel(graphicalLabel: GraphicalLabel, layer: number): Node {
+    public drawLabel(graphicalLabel: GraphicalLabel, layer: number): void {
```
Comment (rvilarl) [inline src/MusicalScore/Graphical/MusicSheetDrawer.ts:146]: I believe that the return value had no use.

File: src/MusicalScore/Graphical/VexFlow/VexFlowConverter.ts [hunk 7]
```diff
@@ -543,10 +541,10 @@ export class VexFlowConverter {
             // add Tremolo strokes (only single note tremolos for now, Vexflow doesn't have beams for two-note tremolos yet)
             const tremoloStrokes: number = notes[i].sourceNote.TremoloStrokes;
             if (tremoloStrokes > 0) {
-                const tremolo: Vex.Flow.Tremolo = new Vex.Flow.Tremolo(tremoloStrokes);
-                (tremolo as any).extra_stroke_scale = rules.TremoloStrokeScale;
-                (tremolo as any).y_spacing_scale = rules.TremoloYSpacingScale;
-                vfnote.addModifier(i, tremolo);
+                const tremolo: Tremolo = new Tremolo(tremoloStrokes);
+                // rvilarl tremolo.extra_stroke_scale = rules.TremoloStrokeScale;
+                // rvilarl tremolo.y_spacing_scale = rules.TremoloYSpacingScale;
```
Comment (rvilarl) [inline src/MusicalScore/Graphical/VexFlow/VexFlowConverter.ts:546]: tremolo does not have these attributes.

File: src/MusicalScore/Graphical/VexFlow/VexFlowMeasure.ts [hunk 8]
```diff
@@ -274,39 +276,43 @@ export class VexFlowMeasure extends GraphicalMeasure {
      */
     public addClefAtEnd(clef: ClefInstruction, visible: boolean = true): void {
         const vfclef: { type: string, size: string, annotation: string } = VexFlowConverter.Clef(clef, "small");
-        if (!visible && (this.stave as any).endClef) {
+        if (!visible && this.stave.getEndClef()) {
             return; // don't overwrite existing clef with invisible clef
         }
         this.stave.setEndClef(vfclef.type, vfclef.size, vfclef.annotation);
+        /* rvilarl see reasons below
```
Comment (rvilarl) [inline src/MusicalScore/Graphical/VexFlow/VexFlowMeasure.ts:283]: This for has no use, I believe.

File: src/MusicalScore/Graphical/VexFlow/VexFlowMeasure.ts [hunk 9]
```diff
@@ -318,10 +324,11 @@ export class VexFlowMeasure extends GraphicalMeasure {
             case SystemLinePosition.MeasureBegin:
                 switch (lineType) {
                     case SystemLinesEnum.BoldThinDots:
+                        /* rvilarl redefition of draw not allowed.
                         //customize the barline draw function if repeat is beginning of system
                         if (!renderInitialLine) {
-                            (this.stave as any).modifiers[0].draw = function(stave: Vex.Flow.Stave): void {
-                                (stave as any).checkContext();
+                            this.stave.getModifiers(0).draw = function(stave: Stave): void {
```
Comment (rvilarl) [inline src/MusicalScore/Graphical/VexFlow/VexFlowMeasure.ts:330]: Probably something to do here.

**Review by rvilarl — COMMENTED** [review]
Comments on code added to help on review.

File: src/MusicalScore/Graphical/VexFlow/VexFlowConverter.ts (hunk 7 above)
Comment (sschmidTU) [reply src/MusicalScore/Graphical/VexFlow/VexFlowConverter.ts:546]: It does in our patched version in `VexFlowPatch/`. We'll need to adjust our VexFlowPatch files in any case.

File: src/MusicalScore/Graphical/VexFlow/VexFlowMeasure.ts (hunk 1 above)
Comment (rvilarl) [reply src/MusicalScore/Graphical/VexFlow/VexFlowMeasure.ts:1254]: Without Vex.Flow, it is not accepted as constructor. Funny.

File: src/MusicalScore/Graphical/VexFlow/VexFlowConverter.ts (hunk 7 above)
Comment (rvilarl) [reply src/MusicalScore/Graphical/VexFlow/VexFlowConverter.ts:546]: I cannot revert, it will not compile.

**Comment by rvilarl** [comment]
I will change to draft to keep the comments. I have made a new PR reverting to old npm packages dependencies.

**Comment by sschmidTU** [comment]
I would have found it more important to remove the refactors changing the return types. Maybe we do want the new webpack etc. to improve the build process.

**Pull request closed without merging**